### PR TITLE
Refactor copy action to v3

### PR DESCRIPTION
**Pull request by mvriel**
I have changed the CopyFile Action to make use of the new Artifact and
Asset abstractions; I had to make a little bit of extra effort to ensure
that it is compatible with previous editions of phpDocumentor. The
CopyFile Action will do a recursive copy of the files reported by
the Assets abstraction.

This also meant that I have expanded the Assets abstraction to include
the concept of a Folder; which means that you will return a collection
of Path objects that you can subsequently retrieve using the Assets
abstraction.

I have explicitly chosen to return Path objects and not complete
Asset objects because that might grow a bit large (being a recursive
action through the folder and all subfolders) and because that may
cause I/O that is never used because the user may not want to
retrieve all assets. In addition, memory would increase significantly
because all files would be stored in memory instead of retrieved
on a per Asset basis.

### CODE REVIEW

File: src/DomainModel/Renderer/Asset/Folder.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor\DomainModel\Renderer\Asset;
+
+use phpDocumentor\DomainModel\Path;
+use Webmozart\Assert\Assert;
+
+class Folder extends \ArrayObject
```
Comment (jaapio) [inline src/DomainModel/Renderer/Asset/Folder.php]: Missing use statement

File: src/Infrastructure/Renderer/FlySystemAssets.php [hunk 2]
```diff
@@ -42,7 +44,25 @@ public function get(Path $location)
...
         }
 
-        return new Asset($this->filesystem->get((string)$location));
+        /** @var File $flyLocation */
+        $flyLocation = $this->filesystem->get((string)$location);
+        if ($flyLocation->isDir()) {
+            $listing = $this->filesystem->listContents((string)$location, true);
+            $locations = array_map(
+                function ($value) {
+                    if ($value['type'] === 'dir') {
```
Comment (jaapio) [inline src/Infrastructure/Renderer/FlySystemAssets.php]: A folder cannot have any child folders?

File: src/Infrastructure/Renderer/FlySystemAssets.php [hunk 3]
```diff
@@ -42,7 +44,25 @@ public function get(Path $location)
...
+
+                    return new Path($value['path']);
+                },
+                $listing
+            );
+
+            return new Asset\Folder($location, array_values(array_filter($locations)));
+        }
+
+        return new Asset($flyLocation->read());
```
Comment (jaapio) [inline src/Infrastructure/Renderer/FlySystemAssets.php]: Is this a potential memory issue? You are reading the full file content here

**Comment by jaapio** [comment]
Instead of creating an asset object for folders and assets we could consider to just create an interface for both. And create an implementation in the infrastructure layer that proxies to flysystem. I can imaging that you want to do this in a future pr.

File: src/DomainModel/Renderer/Assets.php [hunk 4]
```diff
@@ -38,13 +38,14 @@ public function includeTemplateAssets(Template $template);
     public function has(Path $location);
 
     /**
-     * Retrieves the asset with the given name.
+     * Retrieves an array with the requested asset or if the name matches multiple assets (such as with a folder)
+     * then it returns all of them.
      *
```
Comment (mbed67) [inline src/DomainModel/Renderer/Assets.php]: Doesn't it currently return either an Asset (not array of Assets) or a Folder?

**Pull request merged**